src/endless/WMI.cpp

```diff
@@ -150,3 +152,481 @@ BOOL WMI::GetMachineInfo(CString & manufacturer, CString & model)
+    // ... other elements ...
+};
+
+// https://msdn.microsoft.com/en-us/library/windows/desktop/aa362641(v=vs.85).aspx
+typedef enum BcdBootMgrElementTypes {
+    BcdBootMgrObjectList_DisplayOrder = 0x24000001,
+    BcdBootMgrObjectList_BootSequence = 0x24000002,
+    BcdBootMgrObject_DefaultObject = 0x23000003,
+    BcdBootMgrInteger_Timeout = 0x25000004,
+    BcdBootMgrBoolean_AttemptResume = 0x26000005,
+    BcdBootMgrObject_ResumeObject = 0x23000006,
+    BcdBootMgrObjectList_ToolsDisplayOrder = 0x24000010,
+    BcdBootMgrBoolean_DisplayBootMenu = 0x26000020,
+    BcdBootMgrBoolean_NoErrorDisplay = 0x26000021,
+    BcdBootMgrDevice_BcdDevice = 0x21000022,
+    BcdBootMgrString_BcdFilePath = 0x22000023,
+    BcdBootMgrBoolean_ProcessCustomActionsFirst = 0x26000028,
+    BcdBootMgrIntegerList_CustomActionsList = 0x27000030,
+    BcdBootMgrBoolean_PersistBootSequence = 0x26000031
+} BcdBootMgrElementTypes;
+
+
+// Inspired by https://github.com/sysprogs/BazisLib/blob/master/bzshlp/Win32/BCD.cpp and
+// https://social.msdn.microsoft.com/Forums/sqlserver/en-US/a9996e4a-d2d7-4c42-87d2-48096ea47eb5/wmi-bcdobjects-using-c?forum=windowsgeneraldevelopmentissues#cb08edd9-297b-453d-a7ca-6c96574aee3f
+// -- thanks, both.
+class WmiObject {
+public:
+    WmiObject(IWbemServices *pSvc, IWbemClassObject *pClass, IWbemClassObject *pInstance) :
+        m_pSvc(pSvc),
+        m_pClass(pClass),
+        m_pInstance(pInstance)
+    {
+    }
+
+    WmiObject() :
+        m_pSvc(NULL),
+        m_pClass(NULL),
+        m_pInstance(NULL)
+    {
+    }
+
+    HRESULT ExecMethod_4(
+        LPCWSTR wszMethodName,
+        LPCWSTR wszParam1Name, const CComVariant &cvParam1,
+        LPCWSTR wszParam2Name, const CComVariant &cvParam2,
+        LPCWSTR wszParam3Name, const CComVariant &cvParam3,
+        LPCWSTR wszParam4Name, const CComVariant &cvParam4,
+        LPCWSTR wszOutParam1Name = NULL, CComVariant &vOutParam1 = CComVariant());
+
+
+    HRESULT ExecMethod_2(
+        LPCWSTR wszMethodName,
+        LPCWSTR wszParam1Name, const CComVariant &cvParam1,
+        LPCWSTR wszParam2Name, const CComVariant &cvParam2,
+        LPCWSTR wszOutParam1Name = NULL, CComVariant &vOutParam1 = CComVariant())
+    {
+        return ExecMethod_4(wszMethodName,
+            wszParam1Name, cvParam1,
+            wszParam2Name, cvParam2,
+            NULL, CComVariant(),
+            NULL, CComVariant(),
+            wszOutParam1Name, vOutParam1);
+    }
+
+    HRESULT ExecMethod_1(
+        LPCWSTR wszMethodName,
+        LPCWSTR wszParam1Name, const CComVariant &cvParam1,
+        LPCWSTR wszOutParam1Name = NULL, CComVariant &vOutParam1 = CComVariant())
+    {
+        return ExecMethod_4(wszMethodName,
+            wszParam1Name, cvParam1,
+            NULL, CComVariant(),
+            NULL, CComVariant(),
+            NULL, CComVariant(),
+            wszOutParam1Name, vOutParam1);
+    }
+
+    /* If this were a general-purpose library these would be on separate classes. */
```

Comment:
One call I can already think we might want to add: [changing the timeout](https://msdn.microsoft.com/en-us/windows/hardware/drivers/devtest/changing-the-boot-menu-time-out). The default is 30 seconds. As far as I can tell there is no equivalent to `savedefault` so Windows will always be the default selection in the menu. In theory we could write a small boot-time job (for both Windows and Endless OS) that changes the default menu entry. The BCD database is supposedly just a registry hive, and libraries to manipulate Windows registry files on Linux exist.

(Hmm: I wonder if one could have used the registry APIs on Windows rather than this WMI stuff. Probably best to stick with the “documented”, supported API though.)

So, I'll take a look at refactoring this.